### DISCUSSION TITLE
Orthotropic stiffness matrix from elasticity tensor and fillOrthotropic symmetry error

### DISCUSSION THREAD

**Discussion by chakra34** [idaholab/moose#25759] [Q&A Modules: Solid mechanics]
Hi,

I have the following 2 questions regarding Orthotropic elasticity material:

1. I am taking in Young's modulus (E11, E22, E33), Poisson's ratio (nu12, nu23, nu31), and Shear modulus (G12, G23, G31) as input functions (of temperature) and then I fill the stiffness tensor using fill symmetric9 based on the following equation (https://mooseframework.inl.gov/source/materials/ComputeElasticityTensor.html):

![image](https://github.com/idaholab/moose/assets/22983992/08a9aa3e-f28e-4a2b-9eb8-229ce0aecdf1)

In this case, I calculate C11, C22, C33, C12, C13, C23, C44, C55, C66 (from E's, nu's, and G's) and then populate the 4th order stiffness tensor using fillSymmetric symmetric9 option.

Does this make sense ? or is there something I am missing? ( I ran for a quarter cylindrical geometry where I am getting incorrect elastic strains, I should get same values for XX and YY where the orthotropic symmetry is, but I don't get it).

My other question is:
2. If I fill the elastic tensor using the fill method "orthotropic" -- I also take 9 input Young's modulus (E11, E22, E33), Poisson's ratio (nu12, nu23, nu31), and Shear modulus (G12, G23, G31), and calculate nu21, nu32, and nu13 ( using the relation nu21 =  (nu12 * E22)/E11 ) to get the 9 vector. So ideally the following Moose symmetry assertion error should be satisfied:

![image](https://github.com/idaholab/moose/assets/22983992/ea573757-2c44-42f0-96ba-c35bce19156d)

However, I am getting error : 

![image](https://github.com/idaholab/moose/assets/22983992/4bc55f0f-34fb-4e58-948b-153564e9d50a)

As you can see I am getting a difference due to the floating point operations of ~7e-9 and it is triggering the assertion flag. Is there a way to avert this situation ? All my variables have double precision.

Thanks,
Aritra

**Comment by GiudGiud** [comment]
Add more digits?
What error do you get when running the same check in python?

**Reply by chakra34** [reply]
I will check in python, adding more digits, my data is not in engineering notation, as in it is not 1e6, instead it is 1000000.0, if that makes sense

**Reply by GiudGiud** [reply]
adding 0s wont help I think. We re looking for a number you would have truncated here

**Reply by adityakavalur** [reply]
Thanks for taking a look at this @GiudGiud !

I cant comment on the first part of the question, however, for the second part, it looks like the variables being compared have values of e+7, and have a difference of e-9, so the actual order of difference is beyond 15 digits, which is beyond the precision of double correct? I wonder if being able to pass a local tolerance in such cases would help (https://mooseframework.inl.gov/docs/doxygen/moose/namespaceMooseUtils.html#ad9865895ce5bb2c18722efa26d5c7693)

**Reply by GiudGiud** [reply]
ah that could be it

the key is to use a relative tolerance instead of an absolute tolerance

**Reply by chakra34** [reply]
Hi @GiudGiud @adityakavalur , thanks for the replies. I was able to resolve this issue temporarily by scaling the values by 1e-8 so that all the numbers (E,G, nu) are comparable- any suggestion on a better fix? 
Yeah I think we definitely would need to use relative tolerances, but I am not sure if it needs additional code modification for ```RankFourTensor``` or there is a flag we can pass.


However, I do need some suggestion for the first problem:

Basically I get different results when I use fillmethod 'orthotropic' for the stiffness versus if I calculate C11, C12, C13, C22, C23, C33, C44, C55, C66 as functions of E, nu, and G and then use fillmethod 'symmetric 9'. I will look into the exact values of the stiffness tensor from the two cases for my problem to see where the difference lies and update in the thread soon-- but if you have some suggestion, I really appreciate the help.

Thanks again,

-Aritra

**Reply by lynnmunday** [reply]
Can you check that your elastic constants are correct by deforming a single element and make sure it produces the correct stresses for the strains you are putting into it?

**Reply by chakra34** [reply]
yep I will do that and update the thread! thanks a lot @lynnmunday :)

**Reply by lynnmunday** [reply]
Will you share the input file after you do this?  It would be a good test.

**Reply by chakra34** [reply]
yep, sure!

**Reply by chakra34** [reply]
Hi @lynnmunday, so I did a couple of tests to check it out-- 

For a block geometry with a single element both methods (fillmethod orthotropic and symmetric9) gave identical results under uniaxial loading with expected stresses. 

Then I added a temperature field in the block and the two methods gave different results (my Young's modulus and alpha_CTE were not temperature dependent).

![image](https://github.com/idaholab/moose/assets/22983992/d75103df-01cb-4d7d-9255-673beac8b5ba)

following is my code snippet where I directly calculate the orthotropic tensor from 12 elastic constants using fill method ```orthotropic```

```
#include "ComputeVariableOrthotropicElasticityTensor.h"
#include "RotationTensor.h"

registerMooseObject("TensorMechanicsApp", ComputeVariableOrthotropicElasticityTensor);
registerMooseObject("TensorMechanicsApp", ADComputeVariableOrthotropicElasticityTensor);

template <bool is_ad>
InputParameters
ComputeVariableOrthotropicElasticityTensorTempl<is_ad>::validParams()
{
  InputParameters params = ComputeRotatedElasticityTensorBaseTempl<is_ad>::validParams();
  params.addClassDescription("Compute an orthotropic elasticity tensor for elastic constants that "
                             "change as a function of temperature");

  params.addRequiredCoupledVar("temperature", "Coupled temperature");

  params.addRequiredParam<FunctionName>(
      "E11",
      "Function describing the Young's modulus E11");
  params.addRequiredParam<FunctionName>(
      "E22",
      "Function describing the Young's modulus E22");
  params.addRequiredParam<FunctionName>(
      "E33",
      "Function describing the Young's modulus E33");

  params.addRequiredParam<FunctionName>(
      "G12",
      "Function describing the shear modulus G12");
  params.addRequiredParam<FunctionName>(
      "G23",
      "Function describing the shear modulus G23");
  params.addRequiredParam<FunctionName>(
      "G31",
      "Function describing the shear modulus G31");

  params.addRequiredParam<FunctionName>(
      "nu12",
      "Function describing the Poisson's ratio nu12");
  params.addRequiredParam<FunctionName>(
      "nu23",
      "Function describing the Poisson's ratio nu23");
  params.addRequiredParam<FunctionName>(
      "nu13",
      "Function describing the Poisson's ratio nu13");

  return params;
}


template <bool is_ad>
ComputeVariableOrthotropicElasticityTensorTempl<is_ad>::ComputeVariableOrthotropicElasticityTensorTempl(
    const InputParameters & parameters)
  : ComputeRotatedElasticityTensorBaseTempl<is_ad>(parameters),
  _temperature(this->coupledValue("temperature")),
  _temp_name(this->getVar("temperature",0)->name()),
  _E11(this->getFunction("E11")),
  _E22(this->getFunction("E22")),
  _E33(this->getFunction("E33")),
  _G12(this->getFunction("G12")),
  _G23(this->getFunction("G23")),
  _G31(this->getFunction("G31")),
  _nu12(this->getFunction("nu12")),
  _nu23(this->getFunction("nu23")),
  _nu13(this->getFunction("nu13")),
  _delasticity_tensor_dT(
  this->template declarePropertyDerivative<RankFourTensor>(_elasticity_tensor_name, _temp_name) )
{ 
}

template <bool is_ad>
void
ComputeVariableOrthotropicElasticityTensorTempl<is_ad>::computeQpElasticityTensor()
{
  // Get the stiffness values for the current temperature
  current_qp_temperature = 0.0;
  current_qp_temperature = _temperature[_qp];
  _ortho_stiffness_components.resize(12);

  // this should be the order: E11, E22 E33 G12 G23 G31/G13 nu21 nu31 nu32 nu12 nu13 nu23
  // Ansys takes E11 E22 E33 nu12 nu 23 nu13 G12 G23 G13/G31
  E11 = 0.0; E22 = 0.0; E33 = 0.0;
  G12 = 0.0; G23 = 0.0; G31 = 0.0;
  nu12 = 0.0; nu13 = 0.0; nu23 = 0.0;
  nu21 = 0.0; nu31 = 0.0; nu32 = 0.0;

  E11 = _E11.value(current_qp_temperature) * 1e-8 ;
  E22 = _E22.value(current_qp_temperature) * 1e-8 ;
  E33 = _E33.value(current_qp_temperature) * 1e-8 ;

  G12 = _G12.value(current_qp_temperature) * 1e-8 ;
  G23 = _G23.value(current_qp_temperature) * 1e-8 ;
  G31 = _G31.value(current_qp_temperature) * 1e-8 ;

  nu12 = _nu12.value(current_qp_temperature);
  nu13 = _nu13.value(current_qp_temperature);
  nu23 = _nu23.value(current_qp_temperature);

  // _nu21 = _nu12 * (_E22/_E11) ; 
  nu21 = (nu12 * E22) / E11 ;
  // _nu31 = _nu13 * (_E33/_E11) ;   
  nu31 = (nu13 * E33) / E11 ;
  // _nu32 = _nu23 * (_E33/_E22) ; 
  nu32 = (nu23 * E33) / E22 ;


  _ortho_stiffness_components[0]  = E11 ;  
  _ortho_stiffness_components[1]  = E22 ;  
  _ortho_stiffness_components[2]  = E33 ;  
  _ortho_stiffness_components[3]  = G12 ;  
  _ortho_stiffness_components[4]  = G23 ;  
  _ortho_stiffness_components[5]  = G31 ;  
  _ortho_stiffness_components[6]  = nu21 ;  
  _ortho_stiffness_components[7]  = nu31 ;  
  _ortho_stiffness_components[8]  = nu32 ;  
  _ortho_stiffness_components[9]   = nu12 ;  
  _ortho_stiffness_components[10]  = nu13 ;  
  _ortho_stiffness_components[11]  = nu23 ;  


  _ortho_Cijkl.zero();
  _ortho_Cijkl.fillFromInputVector(_ortho_stiffness_components, RankFourTensor:: orthotropic);
  // Assign elasticity tensor at a given quad point
  _elasticity_tensor[_qp] = _ortho_Cijkl *1e8;

  _dortho_Cijkl.zero();
 _delasticity_tensor_dT[_qp] = _dortho_Cijkl;


}

template class ComputeVariableOrthotropicElasticityTensorTempl<false>;
template class ComputeVariableOrthotropicElasticityTensorTempl<true>;
```

I scale the values by 1e-8 to pass the mooseabsoluteFuzzy check

whereas for the fillmethod ``` symmetric9``` the code snippet looks like:

```
#include "ComputeVariableAnisotropicElasticityTensor.h"
#include "RotationTensor.h"

registerMooseObject("TensorMechanicsApp", ComputeVariableAnisotropicElasticityTensor);
registerMooseObject("TensorMechanicsApp", ADComputeVariableAnisotropicElasticityTensor);

template <bool is_ad>
InputParameters
ComputeVariableAnisotropicElasticityTensorTempl<is_ad>::validParams()
{
  InputParameters params = ComputeRotatedElasticityTensorBaseTempl<is_ad>::validParams();
  params.addClassDescription("Compute an anisotropic elasticity tensor (now orthotropic) for elastic constants that "
                             "change as a function of temperature");

  params.addRequiredCoupledVar("temperature", "Coupled temperature");

  params.addRequiredParam<FunctionName>(
      "E11",
      "Function describing the Young's modulus E11");
  params.addRequiredParam<FunctionName>(
      "E22",
      "Function describing the Young's modulus E22");
  params.addRequiredParam<FunctionName>(
      "E33",
      "Function describing the Young's modulus E33");

  params.addRequiredParam<FunctionName>(
      "G12",
      "Function describing the shear modulus G12");
  params.addRequiredParam<FunctionName>(
      "G23",
      "Function describing the shear modulus G23");
  params.addRequiredParam<FunctionName>(
      "G31",
      "Function describing the shear modulus G31");

  params.addRequiredParam<FunctionName>(
      "nu12",
      "Function describing the Poisson's ratio nu12");
  params.addRequiredParam<FunctionName>(
      "nu23",
      "Function describing the Poisson's ratio nu23");
  params.addRequiredParam<FunctionName>(
      "nu13",
      "Function describing the Poisson's ratio nu13");

  return params;
}


template <bool is_ad>
ComputeVariableAnisotropicElasticityTensorTempl<is_ad>::ComputeVariableAnisotropicElasticityTensorTempl(
    const InputParameters & parameters)
  : ComputeRotatedElasticityTensorBaseTempl<is_ad>(parameters),
  _temperature(this->coupledValue("temperature")),
  _temp_name(this->getVar("temperature",0)->name()),
  _E11(this->getFunction("E11")),
  _E22(this->getFunction("E22")),
  _E33(this->getFunction("E33")),
  _G12(this->getFunction("G12")),
  _G23(this->getFunction("G23")),
  _G31(this->getFunction("G31")),
  _nu12(this->getFunction("nu12")),
  _nu23(this->getFunction("nu23")),
  _nu13(this->getFunction("nu13")),
  _delasticity_tensor_dT(
  this->template declarePropertyDerivative<RankFourTensor>(_elasticity_tensor_name, _temp_name) )
{ 
}

template <bool is_ad>
void
ComputeVariableAnisotropicElasticityTensorTempl<is_ad>::computeQpElasticityTensor()
{
  // Get the stiffness values for the current temperature
  current_qp_temperature = 0.0;
  current_qp_temperature = _temperature[_qp];
  _ortho_stiffness_components.resize(9);

  // this should be the order: E11, E22 E33 G12 G23 G31/G13 nu21 nu31 nu32 nu12 nu13 nu23
  // Ansys takes E11 E22 E33 nu12 nu 23 nu13 G12 G23 G13/G31
  E11  = 0.0; E22  = 0.0; E33  = 0.0;
  G12  = 0.0; G23  = 0.0; G31  = 0.0;
  nu12 = 0.0; nu13 = 0.0; nu23 = 0.0;
  nu21 = 0.0; nu31 = 0.0; nu32 = 0.0;

  E11 = _E11.value(current_qp_temperature) ;
  E22 = _E22.value(current_qp_temperature) ;
  E33 = _E33.value(current_qp_temperature) ;

  G12 = _G12.value(current_qp_temperature) ;
  G23 = _G23.value(current_qp_temperature) ;
  G31 = _G31.value(current_qp_temperature) ;

  nu12 = _nu12.value(current_qp_temperature);
  nu13 = _nu13.value(current_qp_temperature);
  nu23 = _nu23.value(current_qp_temperature);

  // _nu21 = _nu12 * (_E22/_E11) ; 
  nu21 = (nu12 * E22) / E11 ;
  // _nu31 = _nu13 * (_E33/_E11) ;   
  nu31 = (nu13 * E33) / E11 ;
  // _nu32 = _nu23 * (_E33/_E22) ; 
  nu32 = (nu23 * E33) / E22 ;

  //  k = 1 - nu12nu21 - nu23nu32 - nu13nu31 - nu12nu23nu31 - nu21nu32nu13 ;  
  k     = 1 - nu12*nu21 - nu23*nu32 - nu13*nu31 - nu12*nu23*nu31 - nu21*nu32*nu13 ;  

  // C11 = E11 * (1 - nu23nu32)/k
  C11 = E11 * ( 1 - nu23*nu32) / k ;

  // C22 = E22 * (1 - nu13nu31)/k
  C22 = E22 * ( 1 - nu13*nu31) / k ;

  // C33 = E33 * (1 - nu12nu21)/k
  C33 = E33 * ( 1 - nu12*nu21) / k ;

  // C12 = E11 * (nu23nu31 + nu21) /k
  C12 = E11 * (nu23*nu31 + nu21) / k ;

  // C13 = E11 * (nu21nu32 + nu31) /k
  C13 = E11 * (nu21*nu32 + nu31) / k ;

  // C23 = E22 * (nu12nu31 + nu32) / k
  C23 = E22 * (nu12*nu31 + nu32) / k ;

  // C44 = G12
  C44 = G12 ;

  // C55 = G31
  C55 = G31 ;

  // C66 = G23
  C66 = G23 ;
 


 // C_ijkl = 'C11 C12 C13 C22 C23 C33 C44 C55 C66'
  _ortho_stiffness_components[0]  = C11 ;  
  _ortho_stiffness_components[1]  = C12 ;  
  _ortho_stiffness_components[2]  = C13 ;  
  _ortho_stiffness_components[3]  = C22 ;  
  _ortho_stiffness_components[4]  = C23 ;  
  _ortho_stiffness_components[5]  = C33 ;  
  _ortho_stiffness_components[6]  = C44 ;  
  _ortho_stiffness_components[7]  = C55 ;  
  _ortho_stiffness_components[8]  = C66 ;  
 
  _ortho_Cijkl.zero();

  _ortho_Cijkl.fillFromInputVector(_ortho_stiffness_components, RankFourTensor:: symmetric9);
  // Assign elasticity tensor at a given quad point
  _elasticity_tensor[_qp] = _ortho_Cijkl;

 _dortho_Cijkl.zero();
 _delasticity_tensor_dT[_qp] = _dortho_Cijkl;


}

template class ComputeVariableAnisotropicElasticityTensorTempl<false>;
template class ComputeVariableAnisotropicElasticityTensorTempl<true>;

```


The input file snippet looks like so:

```

[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 1
  ny = 1
  nz = 1
  xmin = 0.0
  xmax = 1.0
  ymin = 0.0
  ymax = 1.0
  zmin = 0.0
  zmax = 1.0
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]


[Functions]
  [./topfunc]
    type = ParsedFunction
    value = '1e-6*t' # ramped displacement BC
  [../]

  [./temp_function]
    type  = ParsedFunction
    value = '25.0*(y+1.0)'
  [../]

[]

[BCs]

  [./bot_fix_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
  [../]

  [./top_disp_y]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = 'top'
    function = topfunc
  [../]

  [./left_fix_x]
    type = DirichletBC
    variable = disp_x
    boundary = 'left'
    value = 0
  [../]

  [./front_fix_z]
    type = DirichletBC
    variable = disp_z
    boundary = 'front'
    value = 0
  [../]

[]

[Modules/TensorMechanics/Master/All]
  strain = FINITE
  incremental = true
  add_variables = true
  temperature = temp
  eigenstrain_names = eigenstrain
  generate_output = 'stress_xx stress_yy stress_zz stress_xy stress_yz stress_zx'
  use_automatic_differentiation = false
[]
[Materials]

  [thermal_expansion]
    type = ComputeThermalExpansionEigenstrain
    thermal_expansion_coeff = 1.0e-4
    stress_free_temperature = 100.0
    temperature = temp
    eigenstrain_name = eigenstrain
  []

  [elasticity_tensor]
    type = ComputeVariableAnisotropicElasticityTensor or ComputeVariableOrthotropicElasticityTensor
    E11  = 730000000
    E22  = 730000000
    E33  = 2150000000
    G12  = 14165008.63
    G23  = 377271131.9
    G31  = 377271131.9
    nu12 = 0.0595
    nu23 = 0.0847457627118644
    nu13 = 0.0847457627118644
    temperature = temp
  []

  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]


[]

```

Please let me know with any questions.

Also if I want to print out the stiffness tensor for each ```_qp``` should I just use a C++ for loop ? **Also, is there a way to get the 6x6 stiffness matrix that is obtained from the 4x4x4x4 matrices generated by the two methods (this would help me figure out what's the issue).**

Thanks again,

Best,
Aritra

**Comment by lynnmunday** [comment]
@chakra34  For this comment:  "Then I added a temperature field in the block and the two methods gave different results (my Young's modulus and alpha_CTE were not temperature dependent)."

Are you still using a single element and are you getting the same displacement field and strain field?  Are you outputting the stress as a nodal variable.  What stress do you get if you output it as an elemental value and also what are the results still different if the temperature change is constant over the element versus the gradient you have in this example?  I haven't looked at the code and these are just some options I would try to debug this.  

Are those code snippets something you put in or are those the standard fill methods for these classes?  There is a class SymmetricRankFourTensor and some tests that check that it gives the same results as the full tensor that you might be able to use:
/Users/mundlb/projects/moose/modules/tensor_mechanics/test/tests/mandel_notation/symmetric_small_elastic.i

The symmetric tensor can't be rotated because then they would no longer be symmetric.  
@dschwen Do you have anything to add?

**Reply by chakra34** [reply]
Thanks @lynnmunday for the reply.

I am still using a single element for the temperature dependent problem-- and the displacements are different, so it's not just the stress (shown below is the dispX which is different for the 2 cases)

![image](https://github.com/idaholab/moose/assets/22983992/a7b257be-cbc9-4fee-a9a1-70f0748d75ba)


The above code snippets are written by me which extends the orthotropic capability to account for temperature dependency. But I followed the following to compute the stiffness tensor from the Modulus and Poisson's ratio for orthotropic materials:

![image](https://github.com/idaholab/moose/assets/22983992/a279b3c3-5681-49f3-98d3-34b1523d621b)

But am getting the different results, which is the root of my confusion.

I will check the symmetricRankFourTensor tests and see if I can figure out what's the issue.

Thanks again,

Best,
Aritra
